### PR TITLE
Fix problems with timeouts in graphql_transport_ws

### PR DESCRIPTION
**Pull request by kristjanvalur**
This fixes an issue mentioned in #2702

## Description

After merging uniform websocket tests for Starlite extension, sporadic deadlocks were observed.
It turned out that  the timeout trigger, part of the `graphql_transport_ws` protocol, could trigger too early,
when the initial `websockets` handshake was still being done.  This caused the whole websocket connection
attempt to be rejected and this triggered a deadlock in Starlite, which still appears a bit un-robust.

This PR does a few things

1. provides an API to start the timeout, after the integration specific handler has completed the Websockets handshake.
2. Add initial synchronization (`connection_timed_out:bool`) to ensure that there is never a race between a timeout and accepting a connection.
3. Robustly terminate the timeout task when it is done, or no longer needed.
4. Add error handing around the timeout task.  
5. Add a Task error handling hook, to report unhandled errors in Tasks.  This is not yet in use, need to consult on which log handler to use, perhaps "strawberry.task"
6. Provide a shutdown() api to the integrations, rather than the former, rather unwieldly, boilerplate code.

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #2702 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by kristjanvalur** [comment]
Note that I am not exactly sure why the deadlock occurs, it could be a result of the Strawberry integration.  I was uncsuccessful in reproducing it in the plain Starlite repo, and indeed, it contains unittests for websocket rejection using 'close' before an 'accept' message is sent.
But Strawberry always accepts a websocket connection.  The protocols for GraphQL instead provide for a rejection to be done as a result of the sub-protocol handshake.

File: strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py [hunk 1]
```diff
@@ -73,19 +74,51 @@ async def handle_request(self) -> Any:
...
+            self.connection_init_timeout_task.cancel()
+            with suppress(asyncio.CancelledError):
+                await self.connection_init_timeout_task
+
+        for operation_id in list(self.subscriptions.keys()):
+            await self.cleanup_operation(operation_id)
+        await self.reap_completed_tasks()
+
+    def on_request_accepted(self) -> None:
+        #  handle_request should call this once it has sent the
```
Comment (patrick91) [inline strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py:90]:
```suggestion
        # handle_request should call this once it has sent the
```

File: RELEASE.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Release type: patch
+
+`grapql-transport-ws`: Don't start the sub-protocol timeout until the `websockets` connection handshake is complete.
```
Comment (patrick91) [inline RELEASE.md:3]:
```suggestion
This release improves the `graphql-transport-ws` implementation by starting the sub-protocol timeout only when the connection handshake is completed.
```

**Comment by kristjanvalur** [comment]
One question:  
I added a skeleton error handler for background tasks.  The only sensible thing to do is to log them.  but I didn't implement that for this PR.  I could do that now, if we agree on a log channel to use.  I suggested `strawberry.task`, maybe that is ok? 
These are the only tasks created by strawberry, all other top level task logging is done by the web framework in question.

**Comment by patrick91** [comment]
> One question: I added a skeleton error handler for background tasks. The only sensible thing to do is to log them. but I didn't implement that for this PR. I could do that now, if we agree on a log channel to use. I suggested `strawberry.task`, maybe that is ok? These are the only tasks created by strawberry, all other top level task logging is done by the web framework in question.

when are these errors happening?

**Comment by kristjanvalur** [comment]
> when are these errors happening?

Never, hopefully.  But whenever you create a background task, it is prudent to install a top-level error handler to catch and log it.  If you don't, Python will create a warning about an task with an exception not being "awaited" but that warning may end up anywhere.  In case of the timeout thread, there really aren't many things which can go wrong. But for the subscription thread, all kinds of errors can occur and it is best to handle them in-task by a top level error handler.  

The alternative is to have the main task "await" all background tasks and catch and log any errors which occur there.

**Pull request merged**